core/src/test/scala/cats/derived/adtdefns.scala

```diff
@@ -55,8 +55,14 @@ object TestDefns {
   final case class Recursive(i: Int, is: Option[Recursive])
   object Recursive extends ((Int, Option[Recursive]) => Recursive) {
 
-    implicit val arbitrary: Arbitrary[Recursive] =
-      Arbitrary(Gen.delay(Arbitrary.arbitrary[(Int, Option[Recursive])].map(tupled)))
+    implicit val arbitrary: Arbitrary[Recursive] = {
+      def recursive(maxLength: Int): Gen[Recursive] = for {
+        i <- Arbitrary.arbitrary[Int]
+        n <- Gen.choose(0, maxLength - 1)
+        is <- if (n <= 0) Gen.const(None) else Gen.option(recursive(n))
```

Comment:
Yes, this works, but I'd probably just do `recursive(n / 2)` rather than `Gen.choose`. It depends how much nesting you want to test versus how much you want to avoid stackoverflow.

Here's an alternate implementation.

```scala
      def recur(depth: Int): Gen[Recursive] =
        Arbitrary.arbitrary[Int].flatMap { i =>
          if (depth > 0) {
            recur(depth / 2).map(r => Recursive(i, Some(r)))
          } else {
            Gen.const(Recursive(i, None))
          }
        }
      Arbitrary(Gen.sized(sz => recur(sz)))
```

Reply:
Yeah I was thinking whether not to halve the size instead. It should be more robust :+1:

Reply:
But I still want to use `Gen.option` for more randomization.

Reply:
The default size for `Gen.sized` is randomly chosen between 0 and 99.  Yes, you would be getting more randomization, but we're halving with `n / 2` so there are only 8 max depths of `Recursive`.    So maybe `Gen.option` would just distribute the occurrence of the 8 depths more evenly?  It would take experimentation.